rasa/nlu/utils/hugging_face/hf_transformers.py

```diff
@@ -0,0 +1,358 @@
+            )
+            tokens_out += _aligned_tokens
+
+        return tokens_out, token_ids_out
+
+    def _get_token_ids_for_batch(
+        self, batch_examples: List[Message], attribute: Text
+    ) -> Tuple[List[List[Token]], List[List[int]]]:
+
+        batch_token_ids = []
+        batch_tokens = []
+        for example in batch_examples:
+
+            example_tokens, example_token_ids = self._tokenize_example(
+                example, attribute
+            )
+            batch_tokens.append(example_tokens)
+            batch_token_ids.append(example_token_ids)
+
+        return batch_tokens, batch_token_ids
+
+    @staticmethod
+    def _compute_attention_mask(actual_sequence_lengths):
+
+        attention_mask = []
+        max_seq_length = max(actual_sequence_lengths)
+        for index in range(len(actual_sequence_lengths)):
+            example_seq_length = actual_sequence_lengths[index]
+            attention_mask.append(
+                [1] * example_seq_length + [0] * (max_seq_length - example_seq_length)
+            )
+
+        attention_mask = np.array(attention_mask).astype(np.float32)
+
+        return attention_mask
+
+    def _add_padding_to_batch(self, batch_token_ids):
+        padded_token_ids = []
+        # Compute max length across examples
+        max_seq_len = 0
+        actual_sequence_lengths = []
+        for example_token_ids in batch_token_ids:
+            actual_sequence_lengths.append(len(example_token_ids))
+            max_seq_len = max(max_seq_len, len(example_token_ids))
+        # Add padding according to max_seq_len
+        # Some models don't contain pad token, we use unknown token as padding token.This doesn't affect the computation
+        # since we compute an attention mask anyways.
+        # pad_token_id = self.tokenizer.pad_token_id if self.contains_special_token else self.tokenizer.unk_token_id
+        for example_token_ids in batch_token_ids:
+            padded_token_ids.append(
+                example_token_ids
+                + [self.pad_token_id] * (max_seq_len - len(example_token_ids))
+            )
+        return actual_sequence_lengths, padded_token_ids
+
+    @staticmethod
+    def _extract_nonpadded_embeddings(embeddings, actual_sequence_lengths):
+
+        nonpadded_sequence_embeddings = []
+        for index, embedding in enumerate(embeddings):
+            unmasked_embedding = embedding[: actual_sequence_lengths[index]]
+            nonpadded_sequence_embeddings.append(unmasked_embedding)
+
+        return np.array(nonpadded_sequence_embeddings)
+
+    def _compute_batch_sequence_features(self, batch_attention_mask, padded_token_ids):
+
+        model_outputs = self.model(
+            np.array(padded_token_ids), attention_mask=np.array(batch_attention_mask)
+        )
+        sequence_hidden_states = model_outputs[
+            0
+        ]  # sequence hidden states is always the first output from all models
```

Comment:
I think for such a long comment it makes sense to put it on the line before the operation